python/gink/crud.py

```diff
@@ -0,0 +1,149 @@
+"""
+The WSGI API application for gink operations.
+
+To run this application, pass --wsgi gink.crud.app when starting gink.
+
+Set auth key with env AUTH_TOKEN.
+
+"""
+
+from json import loads, dumps
+from os import environ
+
+from .impl.directory import Directory
+
+
+class Crud():
+    """
+    WSGI application to GET, PUT, and DELETE data in a Gink database.
+
+    Example for PUT:
+
+    curl -X PUT http://localhost:8099/key -d 3 -H "Content-Type: application/json"
+
+    # Sets "key" in root directory to 3
+
+    Example for GET:
+
+    curl -X GET http://localhost:8099/key
+
+    # Returns 3
+
+    """
+    def __init__(self):
+        self.root = Directory(arche=True)
+        self.auth_token = environ.get("AUTH_TOKEN")
+
+    def __call__(self, environ, start_response):
+        # If auth token is present, expect token in the Authorization header.
+        if self.auth_token:
+            auth_header = environ.get('HTTP_AUTHORIZATION')
+            if not auth_header or self.auth_token not in auth_header:
+                return self._bad_auth_handler(start_response)
+
+        raw_path = environ.get('PATH_INFO')
+        if not raw_path or raw_path == '/':
+            return self._bad_path_handler(start_response)
+
+        if environ.get("REQUEST_METHOD") == "GET":
+            default = object()
+            result = self.root.get(raw_path.split("/"), default)
+            if result is default:
+                return self._data_not_found_handler(start_response)
+            return self._get_handler(result, start_response)
+
+        elif environ.get("REQUEST_METHOD") == "PUT":
+            request_body: bytes = environ.get("wsgi.input").read()
+            if not request_body:
+                return self._bad_body_handler(start_response)
+
+            content_type = environ.get("HTTP_CONTENT_TYPE")
+            if content_type == "application/octet-stream" or content_type == "application/x-www-form-urlencoded":
+                value = request_body # Default to binary
+            elif content_type == "text/plain":
+                value = request_body.decode()
+            elif content_type == "application/json":
+                value = loads(request_body.decode())
+            else:
+                return self._bad_type_handler(start_response)
+
+            self.root.set(raw_path.split("/"), value)
+            return self._set_handler(start_response)
+
+        elif environ.get("REQUEST_METHOD") == "DELETE":
+            deleted = bool(self.root.delete(raw_path.split("/")))
```

Comment:
The `Directory.delete` method can actually throw a `KeyError` if the path tries to walk into a sub-directory that doesn't exist, so it's probably worth catching that and returning a 404 (same with get/set actually).  Assuming an exception isn't thrown, the `Directory.delete` method will always create a tombstone at the given key, even if there's no value previously at the key, so I don't know that it's worth checking the return value.